src/Symfony/Component/DependencyInjection/Compiler/ResolveNamedArgumentsPass.php

```diff
@@ -0,0 +1,150 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\DependencyInjection\Compiler;
+
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Definition;
+use Symfony\Component\DependencyInjection\Exception\InvalidArgumentException;
+
+/**
+ * Resolves named arguments to their corresponding numeric index.
+ *
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ */
+class ResolveNamedArgumentsPass extends AbstractRecursivePass
+{
+    private $classReflectors = array();
+
+    /**
+     * {@inheritdoc}
+     *
+     * @throws InvalidArgumentException
+     */
+    public function process(ContainerBuilder $container)
+    {
+        $throwingAutoloader = function ($class) { throw new \ReflectionException(sprintf('Class %s does not exist', $class)); };
+        spl_autoload_register($throwingAutoloader);
+
+        try {
+            parent::process($container);
+        } finally {
+            spl_autoload_unregister($throwingAutoloader);
+
+            // Free memory
+            $this->classReflectors = array();
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function processValue($value, $isRoot = false)
+    {
+        if (!$value instanceof Definition) {
+            return parent::processValue($value, $isRoot);
+        }
+
+        $parameterBag = $this->container->getParameterBag();
+
+        if ($class = $value->getClass()) {
+            $class = $parameterBag->resolveValue($class);
+        }
+
+        $calls = $value->getMethodCalls();
+        $calls[] = array('__construct', $value->getArguments());
+
+        foreach ($calls as $i => $call) {
+            list($method, $arguments) = $call;
+            $method = $parameterBag->resolveValue($method);
+            $parameters = null;
+            $resolvedArguments = array();
+
+            foreach ($arguments as $key => $argument) {
+                if (is_int($key) || '' === $key || '$' !== $key[0]) {
+                    if (!is_int($key)) {
+                        @trigger_error(sprintf('Using key "%s" for defining arguments of method "%s" for service "%s" is deprecated since Symfony 3.3 and will throw an exception in 4.0. Use numeric keys or $named arguments instead.', $key, $method, $this->currentId), E_USER_DEPRECATED);
+                    }
+                    $resolvedArguments[] = $argument;
```

Comment:
This is related to my comment in `CheckArgumentsValidityPass`, point **2** about all non-integer, non-$-prefix values already being changed to index values.